File: backend/app/schemas/charts.py
from typing import Literal

from pydantic import BaseModel

Regime = Literal["import", "export"]
Group = Literal["meva", "oziq"]


class MonthlyResponse(BaseModel):
    months: list[str]
    imports: dict[str, list[float | None]]   # year -> [12 значений]
    exports: dict[str, list[float | None]]
    import_grow: dict[str, list[float | None]]
    export_grow: dict[str, list[float | None]]


class GroupTotals(BaseModel):
    total: float
    massa: float


class GroupSummary(BaseModel):
    year: int
    group: Group
    import_: GroupTotals
    export: GroupTotals
    total: GroupTotals


class GroupBreakdownRow(BaseModel):
    name: str
    massa: float
    total: float
    avg: float


class GroupBreakdown(BaseModel):
    year: int
    group: Group
    type: Literal["import", "export", "all"]
    rows: list[GroupBreakdownRow]


class TopItemRow(BaseModel):
    label: str
    value: float


class TopItems(BaseModel):
    year: int
    items: list[TopItemRow]


class RegionPoint(BaseModel):
    name: str         # русское/узбекское имя области (как в БД)
    code: str         # uz-an, uz-tk и т.п.
    value: float
    massa: float
    meva_total: float
    meva_massa: float
    oziq_total: float
    oziq_massa: float


class RegionsResponse(BaseModel):
    year: int
    regime: Regime
    items: list[RegionPoint]
    max_value: float


class WorldPoint(BaseModel):
    iso: str
    name: str
    name_uz: str | None
    value: float
    massa: float
    meva_value: float
    meva_massa: float
    oziq_value: float
    oziq_massa: float


class WorldResponse(BaseModel):
    year: int
    regime: Regime
    items: list[WorldPoint]
    max_value: float

File: frontend/src/types/charts.ts
export type ChartRegime = 'import' | 'export';
export type ChartGroup = 'meva' | 'oziq';

export interface ChartFilters {
  year?: number;
  tnved?: string[];
  region_id?: number;
  country_id?: number;
}

export interface MonthlyResponse {
  months: string[];
  imports: Record<string, Array<number | null>>;
  exports: Record<string, Array<number | null>>;
  import_grow: Record<string, Array<number | null>>;
  export_grow: Record<string, Array<number | null>>;
}

export interface GroupTotals {
  total: number;
  massa: number;
}

export interface GroupSummary {
  year: number;
  group: ChartGroup;
  import_: GroupTotals;
  export: GroupTotals;
  total: GroupTotals;
}

export interface GroupBreakdownRow {
  name: string;
  massa: number;
  total: number;
  avg: number;
}

export interface GroupBreakdown {
  year: number;
  group: ChartGroup;
  type: 'import' | 'export' | 'all';
  rows: GroupBreakdownRow[];
}

export interface TopItemRow {
  label: string;
  value: number;
}

export interface TopItems {
  year: number;
  items: TopItemRow[];
}

export interface RegionPoint {
  name: string;
  code: string;
  value: number;
  massa: number;
  meva_total: number;
  meva_massa: number;
  oziq_total: number;
  oziq_massa: number;
}

export interface RegionsResponse {
  year: number;
  regime: ChartRegime;
  items: RegionPoint[];
  max_value: number;
}

export interface WorldPoint {
  iso: string;
  name: string;
  name_uz: string | null;
  value: number;
  massa: number;
  meva_value: number;
  meva_massa: number;
  oziq_value: number;
  oziq_massa: number;
}

export interface WorldResponse {
  year: number;
  regime: ChartRegime;
  items: WorldPoint[];
  max_value: number;
}
